flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java

```diff
@@ -0,0 +1,429 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.api.operators.sort;
+
+import org.apache.flink.api.common.typeutils.TypeComparator;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.java.functions.KeySelector;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.configuration.AlgorithmOptions;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.core.io.InputStatus;
+import org.apache.flink.core.memory.DataOutputSerializer;
+import org.apache.flink.runtime.checkpoint.channel.ChannelStateWriter;
+import org.apache.flink.runtime.io.AvailabilityProvider;
+import org.apache.flink.runtime.io.disk.iomanager.IOManager;
+import org.apache.flink.runtime.jobgraph.tasks.AbstractInvokable;
+import org.apache.flink.runtime.memory.MemoryAllocationException;
+import org.apache.flink.runtime.memory.MemoryManager;
+import org.apache.flink.runtime.operators.sort.ExternalSorter;
+import org.apache.flink.runtime.operators.sort.PushSorter;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.io.PushingAsyncDataInput;
+import org.apache.flink.streaming.runtime.io.StreamTaskInput;
+import org.apache.flink.streaming.runtime.streamrecord.LatencyMarker;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.streaming.runtime.streamstatus.StreamStatus;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.MutableObjectIterator;
+
+import javax.annotation.Nonnull;
+
+import java.io.IOException;
+import java.util.PriorityQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.IntStream;
+
+/**
+ * An input that wraps an underlying input and sorts the incoming records. It starts emitting records
+ * downstream only when all the other inputs coupled with this {@link MultiInputSortingDataInput} have
+ * finished sorting as well.
+ *
+ * <p>Moreover it will report it is {@link #isAvailable() available} or
+ * {@link #isApproximatelyAvailable() approximately available} if it has some records pending only if the head
+ * of the {@link CommonContext#getQueueOfHeads()} belongs to the input. That way there is only ever one input
+ * that reports it is available.
+ *
+ * <p>The sorter uses binary comparison of keys, which are extracted and serialized when received
+ * from the chained input. Moreover the timestamps of incoming records are used for secondary ordering.
+ * For the comparison it uses either {@link FixedLengthByteKeyComparator} if the length of the
+ * serialized key is constant, or {@link VariableLengthByteKeyComparator} otherwise.
+ *
+ * <p>Watermarks, stream statuses, nor latency markers are propagated downstream as they do not make
+ * sense with buffered records. The input emits a MAX_WATERMARK after all records.
```

Comment:
It emits the maximum watermark it has seen, MAX_WATERMARK could be interpreted as `Long.MAX_VALUE`.